src/graph/impl/KokkosGraph_BFS_impl.hpp

```diff
@@ -0,0 +1,159 @@
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Brian Kelley (bmkelle@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef _KOKKOSGRAPH_BFS_IMPL_HPP
+#define _KOKKOSGRAPH_BFS_IMPL_HPP
+
+#include "Kokkos_Core.hpp"
+#include "KokkosKernels_Utils.hpp"
+#include <vector>
+#include <algorithm>
+
+namespace KokkosGraph {
+namespace Experimental {
+namespace Impl {
+
+template<typename rowmap_t, typename entries_t, typename lno_view_t>
+struct SerialRCM
+{
+  using size_type = typename rowmap_t::non_const_value_type;
+  using lno_t = typename entries_t::non_const_value_type;
+  using host_rowmap_t = Kokkos::View<size_type*, Kokkos::HostSpace>;
+  using host_lno_view_t = Kokkos::View<lno_t*, Kokkos::HostSpace>;
+  using host_lno_2D_view_t = Kokkos::View<lno_t**, Kokkos::LayoutLeft, Kokkos::HostSpace>;
+
+  lno_t numVerts;
+  host_rowmap_t rowmap;
+  host_lno_view_t entries;
+
+  SerialRCM(const rowmap_t& rowmap_, const entries_t& entries_) :
+    numVerts(rowmap_.extent(0) - 1),
+    rowmap(Kokkos::ViewAllocateWithoutInitializing("HostRowmap"), rowmap_.extent(0)),
+    entries(Kokkos::ViewAllocateWithoutInitializing("HostEntries"), entries_.extent(0))
+  {
+    Kokkos::deep_copy(rowmap, rowmap_);
+    Kokkos::deep_copy(entries, entries_);
+  }
+
+  lno_t findPseudoPeripheral()
+  {
+    //Choose vertex with smallest degree
+    lno_t periph = -1;
+    lno_t periphDeg = numVerts;
+    for(lno_t i = 0; i < numVerts; i++)
+    {
+      lno_t deg = rowmap(i + 1) - rowmap(i);
```

Comment:
You could probably return early if you find a node with `deg == 0;`

Reply:
That's a good point. I just added that.